### PR TITLE
Fix nagios parser to parse all perfdata and report info message.

### LINKED ISSUE #5591: [parsers.nagios] Fails to parse extended Nagios output
### Relevant telegraf.conf:

```
[[inputs.exec]]
  commands = ["/bin/bash ./nagios.sh"]

  data_format = "nagios"
```

`nagios.sh`
``` 
#! /bin/bash

# Taken from https://assets.nagios.com/downloads/nagioscore/docs/nagioscore/3/en/pluginapi.html
msg="DISK OK - free space: / 3326 MB (56%); | /=2643MB;5948;5958;0;5968
/ 15272 MB (77%);
/boot 68 MB (69%);
/home 69357 MB (27%);
/var/log 819 MB (84%); | /boot=68MB;88;93;0;98
/home=69357MB;253404;253409;0;253414
/var/log=818MB;970;975;0;980"

echo "$msg"
```

### System info:

telegraf 1.10.0

### Steps to reproduce:

1. Create the config and the script files.
2. Run `telegraf --config ./inputs.exec.conf --test`

nagios,host=localhost,perfdata=/var/log,unit=MB critical_gt=975,critical_lt=0,max=253414,min=0,value=980,warning_gt=970,warning_lt=0 1552677172000000000

### Additional info
- The plugin doesn't parse extended Nagios format properly.
- I would like to have a message reported with the state as well. Shourt message and the long message too.

### PR DESCRIPTION
**Pull request by sergeykhegay**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] ~Associated README.md updated.~ Not needed.
- [x] Has appropriate unit tests.

Example Nagios check output:
```
DISK OK - free space: / 3326 MB (56%); | /=2643MB;5948;5958;0;5968
/ 15272 MB (77%);
/boot 68 MB (69%);
/home 69357 MB (27%);
/var/log 819 MB (84%); | /boot=68MB;88;93;0;98
/home=69357MB;253404;253409;0;253414
/var/log=818MB;970;975;0;980
```

Old behavior:
```
nagios_state,host=localhost state=0i 1552677172000000000
nagios,host=localhost,perfdata=/,unit=MB critical_gt=5958,critical_lt=0,max=5968,min=0,value=2643,warning_gt=5948,warning_lt=0 1552677172000000000
nagios,host=localhost,perfdata=/boot,unit=MB critical_gt=93,critical_lt=0,max=98,min=0,value=68,warning_gt=88,warning_lt=0 1552677172000000000
```

**New** behavior:
```
nagios,host=localhost,perfdata=/,unit=MB critical_gt=5958,critical_lt=0,max=5968,min=0,value=2643,warning_gt=5948,warning_lt=0 1552939211000000000
nagios,host=localhost,perfdata=/boot,unit=MB critical_gt=93,critical_lt=0,max=98,min=0,value=68,warning_gt=88,warning_lt=0 1552939211000000000
nagios,host=localhost,perfdata=/home,unit=MB critical_gt=253409,critical_lt=0,max=253414,min=0,value=69357,warning_gt=253404,warning_lt=0 1552939211000000000
nagios,host=localhost,perfdata=/var/log,unit=MB critical_gt=975,critical_lt=0,max=980,min=0,value=818,warning_gt=970,warning_lt=0 1552939211000000000
nagios_state,host=localhost longmsg="/ 15272 MB (77%); /boot 68 MB (69%); /home 69357 MB (27%); /var/log 819 MB (84%);",msg="DISK OK - free space: / 3326 MB (56%);",state=0i 1552939211000000000
```

Changes:
- Lines for `perfdata` in `{/home, /var/log}` added.
- Info message and long info message are reported for the nagios_state.

Fixes #5591.

### CODE REVIEW

**Comment by sergeykhegay** [comment]
@glinton, could you review please?

**Comment by sergeykhegay** [comment]
Hi @danielnelson , @glinton is there any update on this? Is it being reviewed or any ETA? Regards, Sergey.

**Comment by danielnelson** [comment]
I'll try to do a review later today

File: plugins/parsers/nagios/parser_test.go [hunk 1]
```diff
@@ -1,112 +1,377 @@
...
+					"critical_lt": float64(0),
+					"critical_gt": float64(90),
+					"min":         float64(0),
+					"max":         float64(100),
+				}, metrics[1].Fields())
+
+				assertNagiosState(t, metrics[2], map[string]interface{}{
+					"state":   int64(0),
+					"msg":     "PING OK - Packet loss = 0%, RTA = 0.30 ms",
+					"longmsg": "This is a long output with three lines",
```
Comment (danielnelson) [inline plugins/parsers/nagios/parser_test.go]: Would it make sense to name these: `service_output` and `long_service_output` to match the `$SERVICEOUTPUT$` and `$LONGSERVICEOUTPUT$` macros they correspond to?

Secondly, do you think we should include the `\n` espaced newlines like nagios would?
```
This is a long output\nwith three lines
```

Or in the example from the nagios documentation:
```
/ 15272 MB (77%);\n/boot 68 MB (69%);\n/var/log 819 MB (84%);
```

File: plugins/inputs/exec/exec.go [hunk 2]
```diff
@@ -105,43 +84,46 @@ func (c CommandRunner) Run(
...
 	}
 
 	out = removeCarriageReturns(out)
+
+	if isNagiosParser {
+		exitCode, err := nagios.GetExitCode(nagiosErr)
+		if err != nil {
+			return nil, fmt.Errorf("exec: get exit code: %s", err)
+		}
+		nagios.AppendExitCode(&out, exitCode)
```
[the remark this answers is not in the document]
Comment (sergeykhegay) [reply plugins/inputs/exec/exec.go]: I feel like the `state`, `service_output`, `long_service_output` semantically belong together and should be a part of the same measurement.

The current parser interface doesn't allow to pass any extra information to the parser, whilst the `service_output` and the `long_service_output` are known during the parsing time only.

The parser will properly parse bytes for which the `nagios.AppendExitCode(...)` wasn't called, it will assume the default state to be 0.

File: plugins/parsers/nagios/parser_test.go (hunk 1 above)
Comment (sergeykhegay) [reply plugins/parsers/nagios/parser_test.go]: - I like the `service_output` and `long_service_output`.

- I used the `\n` at first, then I saw https://github.com/influxdata/docs.influxdata.com/pull/2192, which says the line protocol doesn't support the `\n` inside fields. 
Now, that I think about it, the `influxdb` output plugin's work to properly handle this. Since there are other outputs possible, I will use the `\n` again.

**Comment by sergeykhegay** [comment]
Hi @danielnelson, commented and made some changes. Do you think it is good to go now?

File: plugins/inputs/exec/exec.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/inputs/exec/exec.go]: It's not my primary concern but wouldn't the state be set incorrectly depending on how the output ends?  Perhaps we could add the state to the `telegraf.Metric` after it has been returned from the parser?

File: plugins/parsers/nagios/parser_test.go (hunk 1 above)
Comment (danielnelson) [reply plugins/parsers/nagios/parser_test.go]: Yeah, just to clarify those docs the `\n` characters can be in a field but have no special meaning.

File: plugins/inputs/exec/exec.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sergeykhegay) [reply plugins/inputs/exec/exec.go]: - Because we encode the state in the `exec` plugin, no matter how the output ends, the parser will always properly decode the state and strip those additional bytes.

  There might be an issue, when the parser is used somewhere else. And it is my bad, because I did not properly document the behavior of the `Parse` method. I will add the docs, but first, could you please comment on the next point:
- Appending the error code after we return from the parser is definitely possible. But will require me to do something like this:

```golang

// To return the exit code through the error handling mechanism.
//
type exitCode int

func (c exitCode) Error() string {}

// The callee changes
//
func (c CommandRunner) Run(
	e *Exec,
	command string,
	acc telegraf.Accumulator,
) ([]byte, error) {
	<...irrelevant...>

	out = <output from the check>

	if isNagiosParser {
		exitCode, err := nagios.GetExitCode(nagiosErr)
		if err != nil {
			return nil, fmt.Errorf("exec: get exit code: %s", err)
		}
		return out.Bytes(), exitCode(exitCode)
	}

	return out.Bytes(), nil
}

// The caller changes
//
func (e *Exec) ProcessCommand(command string, acc telegraf.Accumulator, wg *sync.WaitGroup) {
	defer wg.Done()

	out, err := e.runner.Run(e, command, acc)
        var exitCode int
	if err != nil {
		if v, ok := err.(exitCode); ok {
			exitCode = int(v)
		} else {
			acc.AddError(err)
			return
		}
	}
      

	metrics, err := e.parser.Parse(out)
	if err != nil {
		acc.AddError(err)
		return
	}
	
        if _, ok := e.parser.(*nagios.NagiosParser); ok {
                // Find the metric with name == "nagios_state" and append the "state" field to it.
                // Can be a convention that it is the last metric, or iterate through all and find it.
               m := <somehow find the metric>
               m.AddField("state", exitCode)
        }
        
        for _, metric := range metrics {
		acc.AddFields(metric.Name(), metric.Fields(), metric.Tags(), metric.Time())
	}
}
```
- I am relying on your expertise here. Which approach seem to be better for you?

[the remark this answers is not in the document]
Comment (sergeykhegay) [reply plugins/inputs/exec/exec.go]: @danielnelson, what is your take on this ^^?

[the remark this answers is not in the document]
Comment (danielnelson) [reply plugins/inputs/exec/exec.go]: To be sure, neither solution is great, and if I could do it over then I would probably make nagios it's own plugin separate from exec and scrap the standalone parser.

Putting that aside though, since I don't think it is worth a breaking change... I would parse, then if this is the NagiosParser, get the error code and iterate over the metrics looking for the `nagios_state` metric using `metric.Name()`.  You will also want to watch for the case where the state metric is not found, and add the field as a new metric.

One more thing, could you do me a favor and use this form for adding the metrics?
```
        for _, metric := range metrics {
		acc.AddMetric(metric)
	}
```

**Comment by sergeykhegay** [comment]
Hi @danielnelson, addressed your comments.

One of the test builds fails though, it seems to be something finicky with the require library. Have you seen something similar before?

There is a new version of it, I'll check if they have related changes there.

**Comment by danielnelson** [comment]
I think the tests fails because the ordering in a map in Go is unspecified, so they can't be safely compared, try using the helper functions in `testutil/metric.go`.

**Comment by sergeykhegay** [comment]
Fixed the tests. But this is very weird. Golang's `reflect.DeepEqual` should have handled this correctly. And the `require` library seems to use it (just skimmed the code).

Please let me know what you think of the changes, @danielnelson.

**Comment by danielnelson** [comment]
Looks great, thanks!
